server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,351 @@
+                return false;
+            }
+            // move below into write lock
+            stats.numRemovalAttempts.inc();
+        } finally {
+            readLock.unlock();
+        }
+        writeLock.lock();
+        try {
+            CounterMetric numCollisions = collidedIntCounters.get(transformedValue);
+            if (numCollisions != null) {
+                // This transformed value has had a collision before
+                HashSet<Integer> removalSet = removalSets.get(transformedValue);
+                if (removalSet == null) {
+                    // First time a removal has been attempted for this transformed value
+                    HashSet<Integer> newRemovalSet = new HashSet<>();
+                    newRemovalSet.add(value); // Add the key value, not the transformed value, to the list of attempted removals for this transformedValue
+                    removalSets.put(transformedValue, newRemovalSet);
+                    numCollisions.dec();
+                } else {
+                    if (removalSet.contains(value)) {
+                        return false; // We have already attempted to remove this value. Do nothing
+                    }
+                    removalSet.add(value);
+                    numCollisions.dec();
+                    // If numCollisions has reached zero, we can safely remove all values in removalList
+                    if (numCollisions.count() == 0) {
+                        removeFromRBM(transformedValue);
+                        collidedIntCounters.remove(transformedValue);
+                        removalSets.remove(transformedValue);
+                        return true;
+                    }
+                }
+                return false;
+            }
+            // Otherwise, there's not been a collision for this transformedValue, so we can safely remove
+            removeFromRBM(transformedValue);
+            return true;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    // Helper fn for remove()
+    private void removeFromRBM(int transformedValue) {
+        rbm.remove(transformedValue);
+        stats.size.dec();
+        stats.numSuccessfulRemovals.inc();
+    }
+
+    @Override
+    public int getSize() {
+        readLock.lock();
+        try {
+            return (int) stats.size.count();
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public int getAddAttempts() {
+        return (int) stats.numAddAttempts.count();
+    }
+
+    @Override
+    public int getCollisions() {
+        return (int) stats.numCollisions.count();
+    }
+
+    @Override
+    public boolean isCollision(Integer value1, Integer value2) {
+        if (value1 == null || value2 == null) {
+            return false;
+        }
+        return transform(value1) == transform(value2);
+    }
+
+     static double getRBMSizeMultiplier(int numEntries, int modulo) {
```

Comment:
numbers are scary, explanation please

Reply:
These are from empirical measurements in a test. I will add a comment explaining it. Is there a good place I can put the writeup for the experiments so everyone can see how it was decided?